File: leetcode/36-valid-sudoku.rb
require 'set'

# @param {Character[][]} board
# @return {Boolean}
def is_valid_sudoku(board)
  board.each_with_index do |row, idx|
    return false unless is_valid_block row
    return false unless is_valid_block (0...9).map {|i| board[i][idx]}
    x = idx / 3
    y = idx % 3
    b3 = (0...9).map do |i|
      x1 = i / 3
      y1 = i % 3
      board[x*3 + x1][y*3 + y1]
    end
    return false unless is_valid_block b3
  end
  true
end

def is_valid_block(block)
  a  = Set.new
  block.each do |b|
    return false if a.include? b
    a.add b if b != '.'
  end
  true
end

File: leetcode/36-valid-sudoku.cpp
/**
 * Determine if a Sudoku is valid, according to: Sudoku Puzzles - The Rules.

 * The Sudoku board could be partially filled, where empty cells are filled with the character '.'.


 * A partially filled sudoku which is valid.

 * Note:
 * A valid Sudoku board (partially filled) is not necessarily solvable. Only the filled cells need to be validated.
 */

class Solution {
public:
    const static int CELL_HEIGHT = 3;

    const static int CELL_WIDTH = 3;

    bool isValidSudoku(vector<vector<char> > &board) {
        for (auto &row : board) {
            unordered_set<char> presented;
            for (auto c : row) {
                if (!testAndSet(c, &presented)) {
                    return false;
                }
            }
        }
        for (auto i(0); i<board[0].size(); ++i) {
            unordered_set<char> presented;
            for (auto j(0); j<board.size(); ++j) {
                if (!testAndSet(board[j][i], &presented)) {
                    return false;
                }
            }
        }
        for (auto i(0); i < board.size() / CELL_HEIGHT; ++i) {
            for (auto j(0); j < board[0].size() / CELL_WIDTH; ++j) {
                unordered_set<char> presented;
                for (auto k(0); k < CELL_WIDTH * CELL_HEIGHT; ++k) {
                    if (!testAndSet(board[i*CELL_HEIGHT + k/CELL_WIDTH][j*CELL_WIDTH + k%CELL_WIDTH], &presented)) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
private:
    bool testAndSet(char c, unordered_set<char> *presented) {
        auto result(false);
        if (c == '.') {
            result = true;
        } else if (presented->find(c) == presented->end()) {
            presented->insert(c);
            result = true;
        }
        return result;
    }
};
